### ISSUE TITLE
Circular flow type definitions trigger no-use-before-define

### ISSUE THREAD

**Issue by qm3ster** [eslint/eslint#8703]
**Tell us about your environment**

* **ESLint Version:** v3.19.0
* **Node Version:** v6.9.5
* **npm Version:** 4.5.0
(but yarn v0.24.5 was used)

**What parser (default, Babel-ESLint, etc.) are you using?**
Babel-ESLint

**Please show your full configuration:**

```
module.exports = {
    extends: [
      "standard",
      "plugin:flowtype/recommended"
   ],
    plugins: [
        "standard",
        "flowtype",
        "promise"
    ]
}
```

**What did you do? Please include the actual source code causing the issue.**
Make a circular type definition in flow

```js
/* @flow */
type File = {chunks: Array<Chunk>, path: string}
type Chunk = {file: File, start: number, end: number, rotation: string}
```
**OR**
```js
/* @flow */
type Chunk = {file: File, start: number, end: number, rotation: string}
type File = {chunks: Array<Chunk>, path: string}
```
**What did you expect to happen?**
Everything to validate

**What actually happened? Please include the actual, raw output from ESLint.**
Errors.
I am sure this can be fixed in `eslint-plugin-flowtype`, but perhaps the fact that rule even tries to apply to this line is a bug? Filing it here because it is a core rule.
```
.../src/commands.js
  2:21  error  'File' was used before it was defined  no-use-before-define

✖ 1 problem (1 error, 0 warnings)
```
**OR**
```
.../src/commands.js
  2:28  error  'Chunk' was used before it was defined  no-use-before-define

✖ 1 problem (1 error, 0 warnings)
```
respectively.

**Comment by soda0289** [comment]
Thanks for the report.

This is an issue with eslint-scope not recognizing type annotations and type parameters. Currently babel-eslint parser will monkey patch the scope analysis tool to allow for these statements to be handled correctly. If I had to guess I think type definitions are not being scoped correctly. If you could open an issue with babel-eslint I can take a look into updating the monkey patch for this scenario.

**Comment by not-an-aardvark** [comment]
Closing because this is an issue with babel-eslint and not with eslint itself.
